### PR TITLE
Remove Foundation dependency

### PR DESCRIPTION
**Pull request by JanStureNielsen**
Remove 'foundation' as a dependency and copy foundation methods into 'nassau' to address #56. If this PR is acceptable, I'll add PRs for the other projects.

### CODE REVIEW

**Comment by jvirtanen** [comment]
@JanStureNielsen I will have a look at this on the weekend. 👍  My apologies for the delay!

File: libraries/core/src/main/java/com/paritytrading/foundation/ASCII.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * Get an ASCII string from a byte array.
+     *
+     * @param bytes a byte array
+     * @return a string
+     */
+    public static String get(byte[] bytes) {
+        return new String(bytes, US_ASCII);
+    }
+
```
[the remark this answers is not in the document]
Comment (JanStureNielsen) [reply libraries/core/src/main/java/com/paritytrading/foundation/ASCII.java]: Now inlined.

File: libraries/core/src/main/java/com/paritytrading/foundation/ByteBuffers.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+     *
+     * @param buffer a buffer
+     * @return the unsigned short value
+     * @throws BufferUnderflowException if there are fewer than two bytes
+     *   remaining in the buffer
+     */
+    public static int getUnsignedShort(ByteBuffer buffer) {
+        return buffer.getShort() & 0xffff;
+    }
+
```
[the remark this answers is not in the document]
Comment (JanStureNielsen) [reply libraries/core/src/main/java/com/paritytrading/foundation/ByteBuffers.java]: Now inlined

File: libraries/core/src/main/java/com/paritytrading/foundation/ByteBuffers.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * @param buffer a buffer
+     * @param value the unsigned short value
+     * @throws BufferOverflowException if there are fewer than two bytes
+     *   remaining in the buffer
+     * @throws ReadOnlyBufferException if the buffer is read-only
+     */
+    public static void putUnsignedShort(ByteBuffer buffer, int value) {
+        buffer.putShort((short)value);
+    }
+
```
[the remark this answers is not in the document]
Comment (JanStureNielsen) [reply libraries/core/src/main/java/com/paritytrading/foundation/ByteBuffers.java]: Now inlined.

File: libraries/core/src/main/java/com/paritytrading/foundation/ByteBuffers.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+     *   written
+     * @param value the unsigned short value
+     * @throws IndexOutOfBoundsException if the index is negative or not
+     *   smaller than the buffer's limit, minus one
+     * @throws ReadOnlyBufferException if the buffer is read-only
+     */
+    public static void putUnsignedShort(ByteBuffer buffer, int index, int value) {
+        buffer.putShort(index, (short)value);
+    }
+
```
[the remark this answers is not in the document]
Comment (JanStureNielsen) [reply libraries/core/src/main/java/com/paritytrading/foundation/ByteBuffers.java]: Now inlined. And ByteBuffers is no longer copied.

File: libraries/core/src/test/java/com/paritytrading/nassau/soupbintcp/SoupBinTCPClientStatus.java [hunk 5]
```diff
@@ -21,8 +22,8 @@ public SoupBinTCPClientStatus() {
 
     @Override
     public void loginAccepted(SoupBinTCPClient client, SoupBinTCP.LoginAccepted payload) {
-        String session        = ASCII.get(payload.session);
-        long   sequenceNumber = ASCII.getLong(payload.sequenceNumber);
+        String session        = new String(payload.session, US_ASCII);
+        long   sequenceNumber = Long.parseLong(new String(payload.sequenceNumber, US_ASCII).trim());
```
Comment (JanStureNielsen) [inline libraries/core/src/test/java/com/paritytrading/nassau/soupbintcp/SoupBinTCPClientStatus.java]: FWIW, I like the `ASCII.get` and `ASCII.getLong` methods better, and I suspect `ASCII.getLong` is more efficient, too (though it probably doesn't matter much).

**Comment by jvirtanen** [comment]
Great! 🙌 

This is looking generally good, and my apologies it has taken a while for me to get to this. Partly that has been because apart from thinking that it would be good to get rid of the Foundation dependency, I hadn't thought too much how things should look like afterwards. 🤡 

The main issue is the `ASCII` class and its usage for the SoupBinTCP messages. I think it will definitely make sense to keep access to the raw binary representation of the fields as is currently done. However, I also think the current API design is fairly bad as it requires one to use `ASCII` (or roll their own implementation) to convert the values to/from `String` and `long`.

I suggest we adjust the API design in the scope of this pull request as follows. Each of the three SoupBinTCP messages (Login Request, Login Rejected, and Login Accepted) would implement convenience methods for getting and setting the fields as `String` or `long`. For example:
```java
/**
 * Set the username to a string.
 *
 * @param username the username
 */
public void setUsername(String username) {
    ASCII.putLeft(this.username, username);
}

/**
 * Get the username as a string.
 *
 * @return the username as a string
 */
public String getUsername() {
    return ASCII.get(username);
}
```
Then `ASCII` can be moved to the `com.paritytrading.nassau.soupbintcp` package and made package-private. Methods that are not needed by the SoupBinTCP support can be removed from that class, as well as Javadoc comments as per the convention.

What do you think?

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Session.java [hunk 6]
```diff
@@ -13,8 +12,6 @@
 
 class Session implements Closeable, MessageListener, SoupBinTCPServerStatusListener {
 
-    private static SoupBinTCP.LoginAccepted loginAccepted = new SoupBinTCP.LoginAccepted();
-
```
Comment (JanStureNielsen) [inline applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Session.java]: Can't multiple `LoginRequest`s be processed a (nearly) the same time?

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Session.java [hunk 7]
```diff
@@ -41,11 +38,7 @@ public void heartbeatTimeout(SoupBinTCPServer session) throws IOException {
 
     @Override
     public void loginRequest(SoupBinTCPServer session, SoupBinTCP.LoginRequest payload) throws IOException {
-        System.arraycopy(payload.requestedSession, 0, loginAccepted.session, 0,
-                payload.requestedSession.length);
-        ASCII.putLongRight(loginAccepted.sequenceNumber, 1);
-
-        downstream.accept(loginAccepted);
+        downstream.accept(new SoupBinTCP.LoginAccepted(payload.getSession(), payload.getSequenceNumber()));
```
Comment (JanStureNielsen) [inline applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Session.java]: This is no longer re-using the static `LoginAccepted` buffers at the cost of a new allocation on each `LoginRequest` -- it seems like this the right thing to do, though...?

File: libraries/core/src/main/java/com/paritytrading/nassau/soupbintcp/SoupBinTCP.java [hunk 8]
```diff
@@ -42,17 +42,24 @@ private SoupBinTCP() {
      * Payload for the Login Accepted packet.
      */
     public static class LoginAccepted {
-        public byte[] session;
-        public byte[] sequenceNumber;
+        private byte[] session;
+        private byte[] sequenceNumber;
```
Comment (JanStureNielsen) [inline libraries/core/src/main/java/com/paritytrading/nassau/soupbintcp/SoupBinTCP.java]: Encapsulate the underlying buffers via constructor initialization and public get methods.

**Comment by JanStureNielsen** [comment]
I think encapsulating the ASCII marshalling within each SoupBinTCP packet is a "Good Thing". This implementation only uses constructor initialization (i.e., no set methods) with get methods. Let me know if we need to change this to match your proposal above.

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Session.java (hunk 6 above)
Comment (jvirtanen) [reply applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Session.java]: No, the gateway uses only a single thread. Only one Login Request will be processed at a time. 🙂

**Comment by jvirtanen** [comment]
Thank you, @JanStureNielsen! 🙌

**Pull request merged**